src/gafaelfawr/providers/oidc.py

```diff
@@ -190,3 +207,49 @@ async def logout(self, session: State) -> None:
             The session state, which contains the GitHub access token.
         """
         pass
+
+    async def _get_ldap_groups(self, uid: str) -> List[TokenGroup]:
+        """Get groups for a user from LDAP.
+
+        Parameters
+        ----------
+        uid : `str`
+            Username of the user.
+
+        Returns
+        -------
+        groups : List[`gafaelfawr.models.token.TokenGroup`]
+            User's groups from LDAP.
+
+        Raises
+        ------
+        gafaelfawr.exceptions.LDAPException
+            One of the groups for the user in LDAP was not valid (missing
+            cn or gidNumber attributes, or gidNumber is not an integer)
+        """
+        assert self._ldap_config
+        group_class = self._ldap_config.group_object_class
+        member_attr = self._ldap_config.group_member
+        ldap_query = f"(&(objectClass={group_class})({member_attr}={uid}))"
+
+        client = bonsai.LDAPClient(self._ldap_config.url)
```

Comment:
So bonasi doesn't need us to manage an app-wide connection pool or anything like that?

Reply:
brianv0 asked about that as well when writing the original, and I decided to use it without a connection pool for now because the long-term goal is to move this out of Gafaelfawr into a separate user information service. It will be a bit slow without a shared connection pool, but since the lookup only happens on initial login, that should be fine for small-scale internal use (at SLAC for instance).

Before we use this for the full user base in conjunction with COmanage and CILogon, I'm going to move the web service responsible for mapping users to group memberships into a separate service and remove stuff like name and email from Gafaelfawr proper, in favor of having it query that service on the fly, for better separation of concerns. That way Gafaelfawr will be purely authentication and authorization, and some other service will be responsible for user metadata and group membership (allowing that service to be swapped out for the needs of a local deployment; some places it may be a simple LDAP gateway, and other places it may do something complicated with COmanage).